Enabled Fn CLI to make calls to OCI API. Updated relevant integration tests.

harshpai opened the pull request:
A pull request for an initial code review of the `Adapter` layer code.

This is the initial code review to validate that the code structure and adapter design used are appropriate. This is a Work In Progress and is not ready for merging in master yet. A final code review will be requested once all implementations and tests are completed.

templecloud commented on adapter/application.go:
```diff
@@ -0,0 +1,35 @@
+	Name string `json:"name,omitempty"`
+
+	// A comma separated list of syslog urls to send all function logs to. supports tls, udp or tcp. e.g. tls://logs.papertrailapp.com:1
+	SyslogURL *string `json:"syslog_url,omitempty"`
```
I think this is the right casing. I remember this use to be SyslogUrl? If not done already might want to double check.

harshpai replied:
// Copied from fn_go/modelsv2/app.go
// We need to maintain this struct since we have already exposed it to the user

templecloud commented on adapter/function.go:
```diff
@@ -0,0 +1,45 @@
+	ID string `json:"id,omitempty"`
+
+	// Hot functions idle timeout before container termination. Value in Seconds.
+	IDLETimeout *int32 `json:"idle_timeout,omitempty"`
```
Just out of interest, why is IDLE capitalized?

harshpai replied:
// Copied from fn_go/modelsv2/fn.go
// We need to maintain this struct since we have already exposed it to the user

templecloud commented on adapter/interface.go:
```diff
@@ -0,0 +1,56 @@
+	DeleteApp(appID string) error
+}
+
+type TriggerClient interface {
```
Intentionally left empty as a marker interface?

templecloud commented on adapter/oci/app_client.go:
```diff
@@ -0,0 +1,74 @@
+}
+
+func (a AppClient) CreateApp(app *adapter.App) (*adapter.App, error) {
+	//TODO: call OCI client
```
Hmmm, this is a bit hard to review without any context. These bits are unimplemented. Is this part of a bigger or multistage strategy for making these changes? If these are really meant to be unimplemented and return null for the short term, then maybe we can some comments explaining the purpose and strategy.

harshpai replied:
This is a WIP. The initial round of code review was to get feedback on class design and adapter code structures. As we move to fully implement the "unimplemented bits" things will get in better shape before the final code review.

templecloud commented on adapter/oss/app_client.go:
```diff
@@ -0,0 +1,103 @@
+
+func (a AppClient) CreateApp(app *adapter.App) (*adapter.App, error) {
+	resp, err := a.client.Apps.CreateApp(&apiapps.CreateAppParams{
+		Context: context.Background(),
```
This looks like a reasonable to create the context.

harshpai replied:
This is just a refactor of the code already present. We are not looking to make changes to or introduce calling cancel, configuring timeouts, etc

templecloud commented on objects/app/apps.go:
```diff
@@ -154,21 +129,21 @@ func CreateApp(a *fnclient.Fn, app *modelsv2.App) (*modelsv2.App, error) {
 
-	fmt.Println("Successfully created app: ", resp.Payload.Name)
-	return resp.Payload, nil
+	fmt.Println("Successfully created app: ", resp.Name)
```
It is probably worth getting the logging story worked out up front. I am new to this CLI code but maybe we could do better? Has this been considered?

harshpai replied:
This is a great point! However, let's keep in mind that this effort is (narrowly) focussed on enabling the CLI to consume OCI APIs instead of OSS APIs. Keeping in mind time constraints it may be better to track the logging improvements in a separate issue.

templecloud commented on adapter/application.go:
```diff
@@ -0,0 +1,35 @@
+package adapter
```
Are there new unit tests for this?

harshpai replied:
There are none as of now. As the implementation is fleshed out, we will work on the testing strategy that may include unit tests.

ostrain commented on commands/deploy.go:
```diff
@@ -44,7 +46,8 @@ func DeployCommand() cli.Command {
 
 type deploycmd struct {
-	clientV2 *v2Client.Fn
+	clientV2      *v2Client.Fn
```
Do we still need clientV2? Does apiClientAdapter replace it?

harshpai replied:
We do need it now since not all references to it have been removed. As the implementation is fleshed out, we will be able to move it from this struct and many others.

harryjsmith replied:
Yes I suppose we'll want to clean this up before this is merged (to keep things clean).

ostrain commented on objects/app/apps.go:
```diff
@@ -22,11 +23,13 @@ import (
-	client   *fnclient.Fn
+	provider         provider.Provider
+	providerAdapter  adapter.Provider
+	client           *fnclient.Fn
```
Same question here, do we need client or can apiClientAdapter replace it?

harshpai replied:
Same answer.

harshpai commented on adapter/oci/app_client.go:
```diff
@@ -0,0 +1,74 @@
+	return resApps, nil
+}
+
+func convertOCIAppsToAdapterApps(ociApps *[]functions.ApplicationSummary) []*adapter.App {
```
Yes, that is correct. I'll get to it when I implement GetApp.

harshpai commented on adapter/oci/fn_client.go:
```diff
@@ -0,0 +1,35 @@
+	return nil
+}
+
+func (f FnClient) ListFn(appID string, limit int64) error {
```
To be implemented.

harshpai commented on adapter/interface.go:
```diff
@@ -0,0 +1,56 @@
+	DeleteApp(appID string) error
+}
+
+type TriggerClient interface {
```
We do know what methods it will have. The same ones used to make manage Triggers in Fn CLI now. This interface will have 2 implementations. An OSS one, that will send calls similar to how we do things now. An OCI one that will be a stub and return an appropriate error response, since we don't support Trigger in OCI Functions.

harshpai commented:
+ Added ability to FN CLI to send calls to OCI Fn Control Plane APIs. 
- Removed the ability of FN CLI to send calls to OSS Fn Control Plane APIs. 
+ Updated relevant integration tests with a new ability to make calls to OCI API.

harryjsmith commented on adapter/oci/app_client.go:
```diff
@@ -0,0 +1,208 @@
+	}
+	req := functions.CreateApplicationRequest{CreateApplicationDetails: body,}
+
+	res, err := a.client.CreateApplication(context.Background(), req)
```
Would there be a benefit in taking a context as a parameter into these methods, rather than using `context.Background()`? (allow caller to set timeout, cancel etc.).

harshpai replied:
We do not provide such facilities (set timeout, cancel etc.) to the caller in our current implementation that consumes Open Source Go CP APIs. Hence, let's keep it the same with the OCI APIs.

harryjsmith commented on adapter/oci/app_client.go:
```diff
@@ -0,0 +1,208 @@
+	//merge config
+	for k, v := range app.Config {
+		if v == "" {
+			delete(app.Config, k)
```
Why do we remove config variables with empty string values here? Is it not valid for a user to set an environment variable with a blank value?

ant1eicher replied:
Yup, it is valid to set a config value to empty string. Tested with OCI CLI:

`{
  "data": {
    "compartment-id": "ocid1.compartment.oc1...",
    "config": {
      "foo": ""
    },
    "defined-tags": {},
    "display-name": "faas-demo",
    "freeform-tags": {},
    "id": "ocid1.fnapp.oc1.iad....",
...`

harryjsmith commented on adapter/oci/app_client.go:
```diff
@@ -0,0 +1,208 @@
+
+func (a AppClient) DeleteApp(appID string) error {
+	req := functions.DeleteApplicationRequest{ApplicationId: &appID,}
+	_, err := a.client.DeleteApplication(context.Background(), req)
```
Should we be checking responses here (and throughout)?

harshpai replied:
Based on the Fn examples in the OCI GO SDK, this is the recommended way to do it.
https://github.com/oracle/oci-go-sdk/blob/master/example/example_functions_test.go#L146

harryjsmith commented on adapter/oci/app_client.go:
```diff
@@ -0,0 +1,208 @@
+}
+
+func convertOCIAppSummaryToAdapterApp(ociApp *functions.ApplicationSummary) *adapter.App {
+	createdAt, _ := strfmt.ParseDateTime(ociApp.TimeCreated.Format(time.RFC3339Nano))
```
Should we be checking and handling `err` here?

harshpai replied:
Nice catch! Fixed it. Thanks.

harryjsmith commented on adapter/oci/fn_client.go:
```diff
@@ -0,0 +1,268 @@
+	//merge config
+	for k, v := range fn.Config {
+		if v == "" {
+			delete(fn.Config, k)
```
Same question as for apps - why do we delete keys with blank values?

harryjsmith commented on adapter/oci/provider.go:
```diff
@@ -0,0 +1,14 @@
+)
+
+type Provider struct {
+	FMCClient *functions.FunctionsManagementClient
```
Very minor nit: Could this just be `FMClient`? The extra `C` seems redundant

harshpai replied:
Fixed.

harryjsmith commented on adapter/application.go:
```diff
@@ -0,0 +1,35 @@
+
+// Copied from fn_go/modelsv2/app.go
+// We need to maintain this struct since we have already exposed it to the user
+type App struct {
```
Thinking out loud here: Is there a reason we want to copy-paste this all, or could we just consume the existing `models` structs from `fn_go`?

harshpai replied:
We would like to be able to eventually retire the `fn_go` library dependency form this Fn CLI. This change allows us to take a step in that direction.

harryjsmith commented on oci-test.sh:
```diff
@@ -0,0 +1,33 @@
+# oci-auth/fn/contexts/functions-test.yaml
+# oci-auth/oci/config
+# simpleapp/app.json
+export FN_API_URL="https://functions.us-ashburn-1.oraclecloud.com/20181201"
```
This domain should be `functions.us-ashburn-1.oci.oraclecloud.com`

harshpai replied:
Will ask the user to fill this in.

harryjsmith commented on test/README.md:
```diff
@@ -37,3 +37,34 @@ The test harness runs a specified CLI  binary(either "../fn" or "TEST_CLI_BINARY
 * Write parallelizable tests: Tests are slow so sequencing them will make the test package slow - the harness includes tools to help make tests isolated (e.g. any app names created with `h.NewFuncName()` will be deleted after a test is done )  - remember to defer `h.Cleanup()` to ensure test state is cleaned up
 * Watch out for the Environment: The CLI package will pass on the surrounding environment to the CLI when its called - (primarily to allow easily overriding local  configuration like FN_API_URL and other env vars)
+
+## Testing against OCI APIs ##
```
Could we test this more tightly, with more mocking rather than relying on having to make actual calls?

harshpai replied:
The current integration tests provide end to end tests against real production OCI API endpoints. This should provide more than sufficient test coverage for this change.

harryjsmith replied:
Yeah, my only concern here is that we can't automate that through CI easily

harshpai commented on adapter/oci/app_client.go:
```diff
@@ -0,0 +1,208 @@
+		return nil, err
+	}
+
+	adapterApp := convertOCIAppToAdapterApp(&res.Application)
```
Based on the Fn examples in the OCI GO SDK, this is the recommended way to do it.
https://github.com/oracle/oci-go-sdk/blob/master/example/example_functions_test.go

harshpai commented on adapter/interface.go:
```diff
@@ -0,0 +1,69 @@
+	return fmt.Sprintf("function %s not found", n.Name)
+}
+
+// NameNotFoundError error for trigger not found when looked up by name
```
Fixed

harshpai commented:
Work In Progress on merge config.

harryjsmith commented:
Superseded by #616
